tests/unit/test_gce_clients.py

```diff
@@ -0,0 +1,344 @@
+    async def test_list_all_items_returns_some_items_on_some_failed_requests(
+            self, mocker, mock_coro, simple_paging_client, caplog):
+        caplog.set_level(logging.ERROR)
+        sleep_mock, sleep_coro = mock_coro
+        mocker.patch('gordon_janitor_gcp.gce_clients.asyncio.sleep', sleep_coro)
+        retries = 3
+        retry_wait = 5
+        with aioresponses() as m:
+            # One successful call
+            m.get(
+                'http://example.com',
+                payload={'data': 'data', 'nextPageToken': 'token'})
+            # second call fails despite retries
+            m.get(
+                'http://example.com/?pageToken=token',
+                exception=exceptions.GCPHTTPError('504 Gateway Timeout'))
+
+            results = await simple_paging_client.list_all_items(
+                'http://example.com', {},
+                retries=retries,
+                retry_wait=retry_wait)
+        assert results == ['data']
+        expected_sleeps = [mocker.call(retry_wait) for _ in range(retries)]
+        sleep_mock.assert_has_calls(expected_sleeps)
+        assert len(caplog.records) == 4
+
+
+class TestCRMClient:
+    @pytest.fixture
+    def crm_base_url(self):
+        return 'https://cloudresourcemanager.googleapis.com/v1/projects?'
+
+    @pytest.fixture
+    def crm_one_page_rsp(self):
+        return {
+            'projects': [
+                {
+                    'projectNumber': '1',
+                    'projectId': 'project-service-1',
+                    'lifecycleState': 'ACTIVE',
+                    'name': 'Project Service 1',
+                    'createTime': '2018-01-01T00:00:00',
+                    'labels': {},
+                    'parent': {}
+                },
+                {
+                    'projectNumber': '2',
+                    'projectId': 'project-service-2',
+                    'lifecycleState': 'ACTIVE',
+                    'name': 'Project Service 2',
+                    'createTime': '2018-02-01T00:00:00',
+                    'labels': {},
+                    'parent': {}
+                }],
+            'nextPageToken': None
+        }
+
+    @pytest.mark.asyncio
+    async def test_list_all_active_projects_correct_params(
+            self, mocker, crm_one_page_rsp, crm_base_url, get_gce_client):
+        crm_client = get_gce_client(gce_clients.CRMClient)
+
+        with aioresponses() as m:
+            m.get(
+                crm_base_url + 'pageSize=500',
+                payload=crm_one_page_rsp)
+
+            results = await crm_client.list_all_active_projects(
+                page_size=500)
+
+        assert results == crm_one_page_rsp['projects']
+
+    @pytest.mark.asyncio
+    async def test_list_all_active_projects_multiple_pages(
+            self, mocker, crm_one_page_rsp, crm_base_url, get_gce_client):
+        crm_client = get_gce_client(gce_clients.CRMClient)
+        page2 = copy.deepcopy(crm_one_page_rsp)
```

Comment:
Is `deepcopy` needed? I thought it was only when objects contain other objects. I think just `crm_one_page_rsp.copy()` would be sufficient.

Reply:
I thought the same, but the nested lists and dicts inside of `crm_one_page_rsp` count as objects, so using shallow copies meant that I was modifying the same dicts.

Reply:
eww okay